File: Solutions/785. Is Graph Bipartite/785. Is Graph Bipartite.cpp
//bfs
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n=graph.size();
        vector<int> colors(n);
        for(int i=0;i<n;++i){
            if(colors[i]!=0){
                continue;
            }
            queue<int> q;
            q.push(i);
            colors[i]=1;
            while(q.size()){
                auto cur=q.front();q.pop();
                for(auto& x:graph[cur]){
                    if(colors[x]==0){
                        colors[x]=-colors[cur];
                        q.push(x);
                    }else if(colors[x]!=-colors[cur]){
                        return false;
                    }
                }
            }
        }
        return true;
    }
};

File: Solutions/785. Is Graph Bipartite/785. Is Graph Bipartite.go


//dfs
func isBipartite(graph [][]int) bool {
    n:=len(graph)
    colors:=make([]int,n)
    for i:=0;i<n;i+=1{
        if colors[i]==0 && !validcolor(&graph,&colors,1,i){
            return false
        }
    }
    return true
}

func validcolor(graph *[][]int,colors *[]int,color,node int) bool{
    if (*colors)[node]!=0{
        return (*colors)[node]==color
    }
    (*colors)[node]=color
    for _,x:=range((*graph)[node]){
        if !validcolor(graph,colors,-color,x){
            return false
        }
    }
    return true
}